### PR TITLE
allow mirror between source control and registry dependencies

### PR DESCRIPTION
**Pull request by tomerd**
motivation: ease the transition to registry based dependencies

changes:
* extend the mirror API to allow mirror between source control and registry dependencies
* make sure target based dependencies take potentially different identity into account
* add tests

### CODE REVIEW

File: Sources/PackageLoading/ManifestJSONParser.swift [hunk 1]
```diff
@@ -97,29 +102,36 @@ enum ManifestJSONParser {
...
                     identityResolver: identityResolver,
-                    fileSystem: fileSystem)
+                    fileSystem: fileSystem
+                )
             } else if type == "registry" {
                 let identity: String = try kindJSON.get("identity")
                 let requirementJSON: JSON = try kindJSON.get("requirement")
                 let requirement = try PackageDependency.Registry.Requirement(v4: requirementJSON)
-                return .registry(identity: .plain(identity), requirement: requirement, productFilter: .everything)
+                //return .registry(identity: .plain(identity), requirement: requirement, productFilter: .everything)
```
Comment (neonichu) [inline Sources/PackageLoading/ManifestJSONParser.swift:129]: leftover?

File: Sources/PackageLoading/ManifestJSONParser.swift [hunk 2]
```diff
@@ -186,8 +199,25 @@ enum ManifestJSONParser {
...
         location = identityResolver.mappedLocation(for: location)
-        // a package in a git location, may be a remote URL or on disk
-        if let localPath = try? AbsolutePath(validating: location) {
+        if PackageIdentity.plain(location).scopeAndName != nil {
+            // re-mapped to registry
+            let identity = PackageIdentity.plain(location)
+            let registryRequirement: PackageDependency.Registry.Requirement
+            switch requirement {
+            case .branch, .revision:
+                throw StringError("invalid mapping of source control to registry, requirement information mismatch.")
```
Comment (neonichu) [inline Sources/PackageLoading/ManifestJSONParser.swift:208]: should we make this more descriptive, e.g. "cannot map branch or revision based dependencies to registry"

**Review by neonichu — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test macOS

**Comment by tomerd** [comment]
@swift-ci smoke test macOS

**Comment by tomerd** [comment]
@swift-ci smoke test macOS

**Comment by tomerd** [comment]
@swift-ci smoke test macOS

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
macOS CI failure related to https://github.com/apple/swift-package-manager/pull/6103

**Comment by MaxDesiatov** [comment]
@swift-ci smoke test macOS

**Comment by tomerd** [comment]
@swift-ci smoke test macOS

**Comment by aschwaighofer** [comment]
@swift-ci clean smoke test macOS

**Comment by tomerd** [comment]
@swift-ci smoke test

**Pull request merged**